### ISSUE TITLE
Economic analysis of block size and block frequency

### ISSUE THREAD

**Issue by bx549** [turtlecoin/meta#109]
TLDR; Any thoughts on the importance of an algorithm/policy for dynamic block size and dynamic block frequency are most welcome. It would be nice to have a protocol that will produce enough revenue to attract a sufficient number of miners while simultaneously keeping tx delays under control.

Background: There have been a few recent articles that predict Bitcoin to become unstable in the post-block reward world when tx fees are the dominant source of revenue for miners. One particular reference is

https://www.cs.princeton.edu/~smattw/CKWN-CCS16.pdf

Another article by Huberman et. al. investigates revenue for miners and tx delay as a function of block size and block frequency (in the context of a post-block-reward world). 

https://moallemi.com/ciamac/papers/bitcoin-2017.pdf

That article is long. Here are my notes/summary of their results:

#### scenario with fixed block size and fixed block creation rate
congestion/delay are necessary to raise revenue for miners.  (even
now, this finding is supported historical data. when blocks are full,
users pay more in tx fees).

large blocks reduce revenue for miners, but delay costs for users
remain (congestion is reduced as more txs get into a block, but there
is still delay because block arrival rate is the same).

more frequent blocks reduce revenue for miners and also reduce delay
costs for users.

#### scenario with dynamic block sizes and block creation rate
large block sizes are bad in that for a particular amount of revenue
(e.g. so that the system has enough miners to be stable) larger block
sizes correspond to increased tx delay (recall that this scenario is
post block reward).

the problem is to find good choices for block size and block creation
rate so that the system is stable (has enough miners) and also tx
delay costs are minimized (i.e. as small as possible).

select smallest block size that is feasible (given technical
considerations) adjust the block creation rate so that congestion
remains constant at the level that generates enough revenue to attract
a sufficient number of miners.

all of the above assumes that tx fees are variable (with some
minimum). another approach is to have an adjustable minimum tx
fee. the paper does not address that problem.

#### So what?

So, I am interested in the economic trade-offs among block size, block
creation rate, tx fees, and (while still in block creation mode) block
reward fees. By "economic trade-offs" I mean sufficient revenue for
miners and palatable delay costs for users. In particular I want to
investigate the idea of an optimal policy for a dynamic block size and
dynamic block creation rate. For most existing protocols that I know
about, the difficulty of mining a block is adjusted so that the block
creation rate stays constant. However, we could investigate the
suggestion by Huberman, that is, adjust the block creation rate so
that congestion remains constant at the level that generates enough
revenue to attract a sufficient number of miners. A natural framework
for determining an optimal policy is Dynamic Programming. Basically
the policy would look something like this: if system state is x, then
block size should be y and block creation rate should be z.

**Comment by bx549** [comment]
Hmmm. I'm not sure how to add an appropriate label. I'm not sure that I have permission to do that.

**Comment by SoreGums** [comment]
so x & y currently exist.  
x is the median tx size.  
y increases based on that and also increases slightly over time as the network height increases.

with the network mixins increasing, the median tx will probably increase as well.

we tested out sending a lot of tx on a testnet recently and managed to get the mempool up to 5,000 tx waiting to get into blocks. once we stopped flooding the mempool the pending tx dropped fairly quickly. the reason why the blocks didn't expand to accommodate more tx is because the signal is tied to the median block size, which didn't increase. the tx were consistently a lot less than the 100KB trigger.

the only real reason for 30sec block times atm is the code currently has enough problems avoiding forks and the emission schedule is tied to the 30sec interval.

ideally block time would be limited by network latency and emission would figure itself out. however the lack of a consistent time factor is a problem that needs to be overcome. right now it is possible to stamp a block with any timestamp as there is not an established way to determine time accurately across the network yet (there are a bunch of considerations required here to avoid forks).

so that is the technical part mostly covered.

as to the economics of the situation, I don't have a coherent position to make on that :)

**Comment by fexra** [comment]
From an economic point of view there are a few things to take into consideration when talking about TRTL:

1) We are very young and most infrastructure isn't there yet
2) We have no real use cases, mass adoption, etc hence no data to analysize
3) TRTL is a multigen coin where we won't hit the the capped supply for a very long time - so this problem is most likely something for future generations to discuss when there is more data  on how the TRTL network is used.

Because of this it is almost impossible to access the economical impact of the current situation and create a mathematical/code solution that will be viable even in a few years down the road with so little current information. We don't even know what the TRTL Network will be in a year because most of it has not been developed yet. Until we have a better picture of our ecosystem with more working code it  and exactly large amounts of users, it will be hard to fully grasp the full economic picture and impacts.

We could create all the models we want, but we first not to have all the variables and understand our ecoystem. We are currently not there at all yet as we just really started on Karia and the core dev is going to get the ball rolling on a core rewrite of the daemons.

"Enough revenue for the miners" is very subjective and almost impossible to predict too with this chaotic market. Profitability will not only rely on the block sizes, times or if they are dynamic or static but also on hundreds of other factors that economists have been trying to model for decades if not centuries. 

Just to give a few that are closer to home: 

- Price of BTC
- BTC dominance
- Perceived Value
- Utility Value
- Demand / Supply
- Liquidity
- Fiat Pairs
- Distribution of Exchanges/Fiat/Crypto
- Economic costs and rewards of the TRTL network
- Financial markets
- Public opinion/branding/marketing
- etc
- etc
- etc

I just think its near impossible to come up with a formula that will guarantee profitability for our miners without taking external data into account like the ones I listed it above. It would be much wiser to just assume that the value of Turtlecoin, which is just the unit of account on the ledger, is directly correlated to the value of the TRTL network. Hence, the more useful the TRTL network is (utility), the more valued Turtlecoin (TRTL) would be as there will be an higher demand for it. This would require Turtlecoin to have intrinsic value and act as a utility.

eg: use TRTL as gas to power a smart contract on the karia sidechain to run an app
eg: use TRTL pay for merchandise online

Of course this is again directly influenced by how we plan and develop our ecosystem.

**Comment by bx549** [comment]
Thanks so much for the comments  ... I was hoping for some insights into what would be useful for the TRTL community. Suppose that you could hire an Economics PhD student this summer and have that student investigate the (future?) costs and rewards of the TRTL network (because I did hire a student and we are looking for a project). I had not thought about the required ingredients for a smart contract network.

**Comment by brandonlehmann** [comment]
Please feel free to reopen this if required.
